spectator-api/src/main/java/com/netflix/spectator/api/ArrayTagSet.java

```diff
@@ -0,0 +1,252 @@
+      if (data.isEmpty()) {
+        return this;
+      } else {
+        Tag[] newTags = new Tag[length + data.size()];
+        System.arraycopy(tags, 0, newTags, 0, length);
+        int i = 0;
+        for (Tag t : data) {
+          newTags[length + i] = BasicTag.convert(t);
+          ++i;
+        }
+        return dedup(newTags);
+      }
+    } else {
+      List<Tag> data = new ArrayList<>();
+      for (Tag t : ts) {
+        data.add(t);
+      }
+      return addAll(data);
+    }
+  }
+
+  /** Add a collection of tags to the set. */
+  ArrayTagSet addAll(Map<String, String> ts) {
+    if (ts.isEmpty()) {
+      return this;
+    } else {
+      Tag[] newTags = new Tag[length + ts.size()];
+      System.arraycopy(tags, 0, newTags, 0, length);
+      int i = length;
+      for (Map.Entry<String, String> entry : ts.entrySet()) {
+        newTags[i++] = new BasicTag(entry.getKey(), entry.getValue());
+      }
+      return dedup(newTags);
+    }
+  }
+
+  /** Add a collection of tags to the set. */
+  ArrayTagSet addAll(String[] ts) {
+    if (ts.length == 0) {
+      return this;
+    } else {
+      int tsLength = ts.length / 2;
+      Tag[] newTags = new Tag[length + tsLength];
+      System.arraycopy(tags, 0, newTags, 0, length);
+      for (int i = 0; i < tsLength; ++i) {
+        newTags[length + i] = new BasicTag(ts[i], ts[i + 1]);
+      }
+      return dedup(newTags);
+    }
+  }
+
+  /** Add a collection of tags to the set. */
+  ArrayTagSet addAll(Tag[] ts) {
+    return addAll(ts, ts.length);
+  }
+
+  /** Add a collection of tags to the set. */
+  ArrayTagSet addAll(Tag[] ts, int tsLength) {
+    if (tsLength == 0) {
+      return this;
+    } else {
+      Tag[] newTags = new Tag[length + tsLength];
+      System.arraycopy(tags, 0, newTags, 0, length);
+      for (int i = 0; i < tsLength; ++i) {
+        newTags[length + i] = BasicTag.convert(ts[i]);
+      }
+      return dedup(newTags);
+    }
+  }
+
+  /**
+   * Dedup any entries in {@code ts} that have the same key. The last entry with a given
+   * key will get selected.
+   */
+  private ArrayTagSet dedup(Tag[] ts) {
+    // It is important for this sort to be stable for the override behavior to work
+    // correctly.
+    Arrays.sort(ts, TAG_COMPARATOR);
+
+    String k = ts[0].key();
+    int j = 0;
+    for (int i = 0; i < ts.length; ++i) {
+      if (k.equals(ts[i].key())) {
+        ts[j] = ts[i];
+      } else {
+        k = ts[i].key();
+        ts[++j] = ts[i];
+      }
+    }
+    return new ArrayTagSet(ts, j + 1);
+  }
+
+  @Override public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+
+    ArrayTagSet other = (ArrayTagSet) o;
+    if (length != other.length) return false;
+
+    for (int i = 0; i < length; ++i) {
+      if (!tags[i].equals(other.tags[i])) return false;
+    }
+    return true;
+  }
+
+  @Override public int hashCode() {
```

Comment:
If not done in id, might be useful to cache so lookups will not recompute each time for a cached id.